include/Components/Foldout.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+** Project FileShare, 2022
+**
+** Author Léo Lhuile
+**
+** Started on  Sat Jan 15 21:08:44 2022 Léo Lhuile
+** Last update Wed Sep 14 22:26:27 2022 Léo Lhuile
+**
+** Foldout.hpp : Class to manage a foldout with icon and extra informations.
+*/
+
+#pragma once
+
+#include <TGUI/TGUI.hpp>
+#include "Components/Button.hpp"
+#include "Components/List.hpp"
+
+namespace FileShare::GUI::Components {
+    class Foldout : public tgui::Group {
+        public:
+            Foldout(const char* typeName = "Components::Foldout", bool initRenderer = true);
+            ~Foldout();
+
+            typedef std::shared_ptr<Foldout> Ptr;
+			typedef std::shared_ptr<const Foldout> ConstPtr;
+
+			static Foldout::Ptr create() { return std::make_shared<Foldout>(); }
+			static Foldout::Ptr copy(Foldout::ConstPtr widget) { return widget ? std::static_pointer_cast<Foldout>(widget->clone()) : nullptr; }
+
+            tgui::Signal &getSignal(tgui::String signalName) override;
+
+            tgui::Signal onClick = { "Clicked" };
+            tgui::Signal onOpen = { "Opened" };
+            tgui::Signal onClose = { "Closed" };
+
+            void addToContent(const tgui::Widget::Ptr &item, const tgui::String &widgetName = "");
+            bool removeFromContent(const Widget::Ptr &widget);
+            const std::vector<tgui::Widget::Ptr> &getContentWidgets() const { return this->content->getWidgets(); }
+
+            void setButtonSize(Components::Button::Size size) { this->button->setSize(size); }
+            Components::Button::Size getButtonSize() const { return this->button->getSize(); }
+
+            void buttonCanBeActive(bool active) { this->button->canBeActive(active, false); }
+            bool buttonCanBeActive() const { return this->button->canBeActive(); }
+
+            void setText(const tgui::String &text) { this->button->setText(text); }
+            const tgui::String &getText() const { return this->button->getText(); }
+
+            void setImage(const tgui::String &image) { this->button->setImage(image); }
+            const tgui::Texture &getImage() const { return this->button->getImage(); }
+
+            void setImageSize(float size) { this->button->setImageSize(size); }
+            void setImageSize(const tgui::Vector2f &size) { this->button->setImageSize(size); }
+            tgui::Vector2f getImageSize() const { return this->button->getImageSize(); }
```

Comment:
Instead of all theses functions, how about : 
```c++
  Button &getButton();
  const Button &getButton() const;
```
Calling code can do `foldout.getButton().setText("ABC");` instead of `foldout.setButtonText("ABC");` which is almost the same thing, but greatly reduces the amount of functions we have to write.

Furthermore, if TGui adds more functions to buttons, or we need to access a function of button we haven't exposed yet, we will have to add it manually. It increase maintenance by a lot.

Reply:
It was to create an abstraction, to separate the fact that the Foldout class uses the Button class.
And there won't be any more functions normally, I didn't include everything that was present in the Button class but only what can be configured for the Foldout class.

I assumed that if a foldout is created, we only have access to these functions in order not to modify the behavior of the button in the foldout, only the foldout can be manipulated this way.

Reply:
They are wrappers, indeed, but they are important to properly separate the logic of the Button and the Foldout.
Moreover, if I leave full access to all the Button functions, nothing will prevent me from calling a Button function that would go against the Foldout, which would cause unexpected behavior and potentially complicated issues to resolve.

I see it more as a security measure rather than a simple wrapper.
Additionally, they are inline by default (if I recall correctly from the compilation behavior? And if not, I can add the `inline` keyword), so in terms of execution, it shouldn't change much.

If one day I really have too many functions to add concerning the Button, I can review this part of the code then. But currently, just for code safety, I would prefer to leave it as it is

Reply:
I dont think we need the security : we can be responsible and not use any function that would break the behavior of the foldout

Yeah they are inline since definition is in hpp, my concern wasn't performance, but maintainability.

We can keep them for now

Reply:
I dont think we need the security : we can be responsible and not use any function that would break the behavior of the foldout

Yeah they are inline since definition is in hpp, my concern wasn't performance, but maintainability.

We can keep them for now